### PR TITLE
Porting legacy reflection_pad2d to ATen

### PR DESCRIPTION
**Pull request by mrshenli**
1. Avoided using `THCDeviceTensor` by re-calculating the mapping from cuda (blockIdx, threadIdx) to input/output tensor index. 
2. Changed Camelcase naming to underscore naming.

@gchanan 

Benchmark:

screen shot 2018-12-24 at 5 57 04 pm

screen shot 2018-12-24 at 5 42 56 pm

### CODE REVIEW

File: aten/src/ATen/native/ReflectionPadding2d.cpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+  AT_CHECK(output_h == grad_output_.size(dim_h),
+    "gradOutput height unexpected. Expected: ", output_h, ", Got: ",
+    grad_output_.size(dim_h));
+
+  /* get contiguous gradOutput */
+  Tensor grad_output = grad_output_.contiguous();
+
+  /* backprop */
+  if (input.ndimension() == 3) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```
Comment (gchanan) [inline aten/src/ATen/native/ReflectionPadding2d.cpp]: I don't think this should dispatch on half.

File: aten/src/ATen/native/ReflectionPadding2d.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+          nplane,
+          input_w, input_h, output_w, output_h,
+          pad_l, pad_t);
+      }
+    );
+  } else {
+    int64_t p;
+#pragma omp parallel for private(p)
+    for (p = 0; p < nbatch; p++) {
+      AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```
Comment (gchanan) [inline aten/src/ATen/native/ReflectionPadding2d.cpp]: same here.

File: aten/src/ATen/native/ReflectionPadding2d.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  reflection_pad2d_out_template(output, input, padding);
+  return output;
+}
+
+Tensor& reflection_pad2d_backward_out_cpu(
+    Tensor& grad_input,
+    const Tensor& grad_output,
+    const Tensor& input,
+    IntList padding) {
+  grad_input = at::zeros_like(input);
```
Comment (gchanan) [inline aten/src/ATen/native/ReflectionPadding2d.cpp]: as mentioned in the other PR, zeros_like isn't correct.

File: aten/src/ATen/native/cuda/ReflectionPadding2d.cu [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  auto output = at::empty({0}, input.options());
+  reflection_pad2d_out_template(output, input, padding);
+  return output;
+}
+
+Tensor& reflection_pad2d_backward_out_cuda(
+    Tensor& grad_input, const Tensor& grad_output,
+    const Tensor& input,
+    IntList padding) {
+  grad_input = at::zeros_like(input);
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/ReflectionPadding2d.cu]: same here.

**Review by gchanan — COMMENTED** [review]
looks good, same comments apply as in https://github.com/pytorch/pytorch/pull/15480.

File: aten/src/ATen/native/ReflectionPadding2d.cpp (hunk 1 above)
Comment (mrshenli) [reply aten/src/ATen/native/ReflectionPadding2d.cpp]: Hi @gchanan, I want to make sure I understand this correctly. Is it correct that all backward functions do not need to be dispatched to half, but forward functions DO need to be dispatched to half. 

I tried remove half for forward pass as well, but hitting the following error:

```
py.test test/test_nn.py -k ReflectionPad -v -s
...                                    
test/test_nn.py::TestNN::test_ReflectionPad2d FAILED                                                                                       
...
RuntimeError: "reflection_pad2d_out_template" not implemented for 'torch.cuda.HalfTensor'
```

**Comment by mrshenli** [comment]
now forward

```python
In [24]: x = torch.rand(80, 80, 80, 80)

In [25]: %timeit torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)
32.6 ms ± 786 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [26]: x = torch.rand([80, 80, 80, 80], device='cuda')

In [27]: %timeit torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)
3.32 ms ± 1.44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

now backward

```python
In [32]: x = torch.rand(80, 80, 80, 80)

In [33]: x_grad = x.requires_grad_()

In [34]: y = torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)

In [35]: grad = torch.randn_like(y)

In [36]: %timeit y.backward(grad, retain_graph=True)
45.1 ms ± 858 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [38]: x = torch.rand([80, 80, 80, 80], device='cuda')

In [39]: x_grad = x.requires_grad_()

In [40]: y = torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)

In [41]: grad = torch.randn_like(y)

In [42]: %timeit y.backward(grad, retain_graph=True)
6.24 ms ± 3.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

before forward

```python
In [12]: x = torch.rand(80, 80, 80, 80)

In [13]: %timeit torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)
32 ms ± 1.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [14]: x = torch.rand([80, 80, 80, 80], device='cuda')

In [15]: %timeit torch.nn.ReflectionPad2d([3, 3, 2, 2])(x)

2 ms ± 1.36 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

before backward

```python
In [17]: x = torch.rand(80, 80, 80, 80)

In [21]: %timeit y.backward(grad, retain_graph=True)
171 ms ± 32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [22]: x = torch.rand([80, 80, 80, 80], device='cuda')

In [26]: %timeit y.backward(grad, retain_graph=True)
5.24 ms ± 4.73 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by mrshenli** [comment]
@gchanan Is there a way to just rerun a single failed test? I do not have the access to click the `rerun workflow` button in ci console.

**Review by gchanan — CHANGES_REQUESTED** [review]

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**